compiler/rustc_hir_analysis/src/hir_ty_lowering/dyn_compatibility.rs

```diff
@@ -352,6 +353,16 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                     b.projection_term = replace_dummy_self_with_error(tcx, b.projection_term, guar);
                 }
 
+                if let Some(term_ty) = &b.term.as_type() {
```

Comment:
Could you actually move *this* check and error reporting into the `for` loop on line 116? That's where we are doing other type alias expansion validation, so I think it's more appropriate for the error reporting to live there.